modules/compression/src/main/java/org/apache/ignite/spi/communication/DeflaterEngine.java

```diff
@@ -0,0 +1,189 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.spi.communication;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.zip.DataFormatException;
+import java.util.zip.Deflater;
+import java.util.zip.Inflater;
+
+import org.apache.ignite.internal.util.nio.compression.CompressionEngine;
+import org.apache.ignite.internal.util.nio.compression.CompressionEngineResult;
+import org.apache.ignite.internal.util.typedef.internal.U;
+
+import static org.apache.ignite.internal.util.nio.compression.CompressionEngineResult.BUFFER_OVERFLOW;
+import static org.apache.ignite.internal.util.nio.compression.CompressionEngineResult.BUFFER_UNDERFLOW;
+import static org.apache.ignite.internal.util.nio.compression.CompressionEngineResult.OK;
+
+/**
+ * Implementation of Deflater algorithm.
+ */
+public final class DeflaterEngine implements CompressionEngine {
+    /** */
+    private static final int initArrSize = 1 << 15;
+
+    /** */
+    private static final int batchSize = 1 << 10;
+
+    /** */
+    private final Deflater deflater = new Deflater();
+
+    /** */
+    private byte[] inputDeflaterArr = new byte[initArrSize];
+
+    /** */
+    private final ExtendedByteArrayOutputStream deflateBaos = new ExtendedByteArrayOutputStream(initArrSize);
+
+    /** */
+    private final Inflater inflater = new Inflater();
+
+    /** */
+    private byte[] inputInflaterArr = new byte[initArrSize];
+
+    /** */
+    private final ExtendedByteArrayOutputStream inflateBaos = new ExtendedByteArrayOutputStream(initArrSize);
+
+    /** */
+    private int inputUnwrapPos = 0;
+
+    /** */
+    private int inputUnwrapLen = 0;
+
+    /** */
+    public DeflaterEngine(){
+        deflater.setLevel(Deflater.BEST_SPEED);
+    }
+
+    /** {@inheritDoc} */
+    @Override public CompressionEngineResult compress(ByteBuffer src, ByteBuffer buf) {
+        assert src != null;
+        assert buf != null;
+
+        int len = src.remaining();
+
+        if (inputDeflaterArr.length < len)
+            inputDeflaterArr = new byte[U.ceilPow2(inputDeflaterArr.length)];
+
+        src.get(inputDeflaterArr, 0, len);
+
+        deflater.reset();
+        deflater.setInput(inputDeflaterArr, 0, len);
+        deflater.finish();
+
+        deflateBaos.reset();
+
+        byte[] arr = new byte[batchSize];
+
+        while (!deflater.finished()) {
+            int cnt = deflater.deflate(arr);
+
+            deflateBaos.write(arr, 0, cnt);
```

Comment:
Maybe better to split it in one line? like `deflateBaos.write(arr, 0, deflater.deflate(arr));` ?

Reply:
Done.